src/pythoninlua.c

```diff
@@ -183,8 +184,52 @@ static int py_object_call(lua_State *L)
         ret = py_convert(L, value);
         Py_DECREF(value);
     } else {
-        PyErr_Print();
-        luaL_error(L, "error calling python function");
+        char s_exc[1024] = {0};
+        char s_traceback[1280] = {0};
+
+        PyObject *exc_type, *exc_value, *exc_traceback;
+        PyErr_Fetch(&exc_type, &exc_value, &exc_traceback);
+        PyErr_NormalizeException(&exc_type, &exc_value, &exc_traceback);
+
+        PyObject *exc_str = PyObject_Str(exc_value);
+
+        // Need not be garbage collected as per documentation of PyUnicode_AsUTF8
+        const char *exc_cstr = (exc_str)?PyUnicode_AsUTF8(exc_str):"";
+        strncpy(s_exc, (!(exc_cstr)?"UNKNOWN ERROR":exc_cstr), 1023);
+
+        if (exc_value != NULL && exc_traceback != NULL) {
+            PyObject *traceback_module = PyImport_ImportModule("traceback");
+            if (traceback_module != NULL) {
+                PyObject *traceback_list = PyObject_CallMethod(traceback_module,
+                        "format_exception", "OOO", exc_type, exc_value, exc_traceback);
+                if (traceback_list != NULL) {
+                    PyObject *traceback_str = PyUnicode_Join(PyUnicode_FromString(""), traceback_list);
+                    if (traceback_str != NULL) {
+                        // Need not be garbage collected as per documentation of PyUnicode_AsUTF8
+                        const char *traceback_cstr = PyUnicode_AsUTF8(traceback_str);
+                        if (traceback_cstr != NULL) {
+                            strncpy(s_traceback, traceback_cstr, 1023);
+                        }
+                        Py_XDECREF(traceback_str);
+                    }
+                    Py_XDECREF(traceback_list);
+                }
+                Py_XDECREF(traceback_module);
+            }
+        }
+
+        if (*s_traceback == '\0') {
+            strcpy(s_traceback, "Exception: ");
+            strncat(s_traceback, s_exc, (1280 - strlen("Exception: ")));
+        }
```

Comment:
Resolved this by avoiding additional string formatting as below
```
if (*s_traceback == '\0') {
    luaL_error(L, "error calling python function:\nException: %s", s_exc);
}
else {
    luaL_error(L, "error calling python function:\nException: %s", s_traceback);
}
```

Added test in test_py.lua for this.

Reply:
You should be able to replace it with `lua_pushfstring`, Something like:

```
if (*s_traceback == '\0')
{
    lua_pushfstring(L, "error calling python function:\nException: %s", s_exc);
}
else
{
    lua_pushfstring(L, "error calling python function:\nException: %s", s_traceback);
}
```

and then at the bottom after all those `Py_DECREF`, change `luaL_error` to `lua_error` instead. This will use the error string at the top of lua's virtual stack.

Reply:
> all those `Py_XDECREF()` at the bottom doesn't happen

The set of ```Py_XDECREF``` invocations are before ```luaL_error```. So management of memory would have taken place.
(Freeing of memory happens after the reference count becomes 0)

This is the reason a copy of the exception traceback/message is being kept in local stack variables ```s_traceback and s_exc```, so that any unintentional reference to freed memory is not there.

Reply:
Maybe I'm missing something here but this segment turns into unreachable code
```
        Py_XDECREF(exc_type);
        Py_XDECREF(exc_value);
        Py_XDECREF(exc_traceback);
        Py_XDECREF(exc_str);

        luaL_error(L, "error calling python function:\n%s", s_traceback);
```
because of this segment above it

```
if (*s_traceback == '\0')
{
  luaL_error(L, "error calling python function:\nException: %s", s_exc);
}
else
{
  luaL_error(L, "error calling python function:\nException: %s", s_traceback);
}
```

Reply:
[The code](https://github.com/bastibe/lunatic-python/blob/e6837bf1fd1f4b75c9e069f497cf7f5029e835b3/src/pythoninlua.c#L200) looks as below, ```luaL_error``` is after ```Py_XDECREF```. Will this become an issue

```
        if (exc_value != NULL && exc_traceback != NULL) {
            PyObject *traceback_module = PyImport_ImportModule("traceback");
            if (traceback_module != NULL) {
                PyObject *traceback_list = PyObject_CallMethod(traceback_module,
                        "format_exception", "OOO", exc_type, exc_value, exc_traceback);
                if (traceback_list != NULL) {
                    PyObject *traceback_str = PyUnicode_Join(PyUnicode_FromString(""), traceback_list);
                    if (traceback_str != NULL) {
                        // Need not be garbage collected as per documentation of PyUnicode_AsUTF8
                        const char *traceback_cstr = PyUnicode_AsUTF8(traceback_str);
                        if (traceback_cstr != NULL) {
                            strncpy(s_traceback, traceback_cstr, 1023);
                        }
                        Py_XDECREF(traceback_str);
                    }
                    Py_XDECREF(traceback_list);
                }
                Py_XDECREF(traceback_module);
            }
        }
        Py_XDECREF(exc_type);
        Py_XDECREF(exc_value);
        Py_XDECREF(exc_traceback);
        Py_XDECREF(exc_str);

        if (*s_traceback == '\0') {
            luaL_error(L, "error calling python function:\nException: %s", s_exc);
        }
        else {
            luaL_error(L, "error calling python function:\nException: %s", s_traceback);
        }
```

Reply:
ah okie I see it now, for some reason github made it appear like the old code was still there.

Reply:
One thought for the refactoring in the future (for python exceptions) that you mentioned in the prior comment.

If it becomes necessary to do the extraction in more than one place, this piece of code can be carved into a separate local function within ```pythoninlua.c``` and invoke it from different places